### PR TITLE
Add Enterprise CommCare Client Version Compliance Report

### PR DESCRIPTION
**Pull request by jingcheng16**
## Product Description

New Tile
image

Report
image

API response
image

## Technical Summary

Ticket: https://dimagi.atlassian.net/browse/SAAS-16236

I'm a bit worried about the performance. When I test on production, it can take more than 2 minutes to generate total number for account like CRS

## Feature Flag

Doing iterative release, so no FF. Once it is merged, we release it.

### Safety story

Tested locally, on staging, and also on production.

### Automated test coverage

Add test for util function I wrote

### QA Plan

QA Ticket: https://dimagi.atlassian.net/browse/QA-7315

Verify the tile works as expected:

- An enterprise account without any mobile workers should see '--' shown up in tile.

Email the report:
- A mobile worker doesn't have any device yet, and haven't submitted any forms yet, will not be included in the report
- A mobile worker whose latest submission is using the latest commcare version, will **not be included** in the report. 

- A mobile worker whose latest submission is using an older commcare version, ([instruction to install an older version](https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/2143946363/Offline+Install+of+CommCare+on+Android)), this mobile worker **will be included** in the report. Then use the same mobile worker, but submitted a form using the "Log in as xx mobile worker" feature, verify this mobile worker **is not** in the report. Because I assume in this case we're always using the latest commcare version. (Using the same mobile worker is to verify the submission through "Log in as xx mobile worker" feature will replace the previous submission when calculate version number)

- Can a mobile worker submitted form using script? If yes, please verify if the latest submission is by script, this mobile worker **is not** in the report. Because I assume in this case we're always using the latest commcare version.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/builds/utils.py
Comment (mjriley) [inline corehq/apps/builds/utils.py:50]: change to: return the latest version available now

File: corehq/apps/builds/utils.py
Comment (mjriley) [inline corehq/apps/builds/utils.py:78]: Nit: move beneath the function that uses it

File: corehq/apps/builds/tests/test_utils.py
Comment (mjriley) [inline corehq/apps/builds/tests/test_utils.py:25]: Test can be removed

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:10]: It's a smell to import a private function. Should this now be a public function?

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:395]: change to lazy translation

Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:395]: Metric may be confusing as it may suggest Kissmetrics or Google Analytics. We may want to rename it? Description, subtitle, etc.

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:407]: Nit: remove space

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:417]: Nit: tab formatting

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:419]: Possibly could be turned into its own function, like: `get_last_version_for_user(user)`. This could be especially important if this same logic is used in the Application Status Report.

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:457]: Change to 1 decimal place

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:453]: There is definitely a performance concern here, as if this summary query takes 2 minutes, that is 2 minutes blocking a web worker. It's worth trying to see if Elasticsearch is any faster here. Otherwise, we may want to cache this statistic.

File: corehq/apps/enterprise/static/enterprise/js/project_dashboard.js
Comment (mjriley) [inline corehq/apps/enterprise/static/enterprise/js/project_dashboard.js:182]: If this is intended to support percentages, do we need to be concerned about locale formatting (i.e. using ',' for a decimale separator) for those as well?

**Review by mjriley — CHANGES_REQUESTED** [review]
Most of the comments are non-blocking, but the most serious was just the performance concern for the summary tile. If Elasticsearch turns out to not be an option, or if that option is still too slow, we'd like to change this so that the data is cached, with the theory that we can stomach blocking our webworkers for 2 minutes every once in a while, but don't want to do so repeatedly.

**Comment by mjriley** [comment]
Instead of caching (which still could generate some negative effects if multiple enterprises all hit the enterprise console at the same time), we could also look at changing the system entirely. Right now, the enterprise console sends out a request for each tile and essentially blocks until the webworker processes it. Instead, we could send an initial request to the webworker which offloaded the task to celery, and the page would then poll the webworker to see if the celery task had completed. Not something we need to do here, but its an option in general for this and other tiles

File: corehq/apps/builds/utils.py
Comment (jingcheng16) [reply corehq/apps/builds/utils.py]: a65c782bd982626d2c8526be4125bde6a5b0d26b

File: corehq/apps/builds/utils.py
Comment (jingcheng16) [reply corehq/apps/builds/utils.py:86]: a44b2002dac682103c2027238e0a5111ca72427d

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: a65c782bd982626d2c8526be4125bde6a5b0d26b

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: a65c782bd982626d2c8526be4125bde6a5b0d26b

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: 3016d9e0505c907532e77a036bec54e81afeb813

File: corehq/apps/builds/tests/test_utils.py
Comment (jingcheng16) [reply corehq/apps/builds/tests/test_utils.py]: 3016d9e0505c907532e77a036bec54e81afeb813

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: c25a980ec516215fb65b406299624c21a66dbd46

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: 5f91e2b1109de4e49d6cd9cb41b39fff2d60ed0e

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: c9586dc2534329070e385a95e7bd1914a8b379a7
I also make it a function to format the percentage, so both you and I can reuse it, thus making sure the percentage will always have just one digit.

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:452]: 1700c2f6d664f5435b938324e2ca24365ac560fc

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: 2e874e57c19bd581866277a544408ff90c7bc6a7

File: corehq/apps/enterprise/static/enterprise/js/project_dashboard.js
Comment (jingcheng16) [reply corehq/apps/enterprise/static/enterprise/js/project_dashboard.js]: Cherry picked your commit in PR #35480 in 1b67a86b55f8e73d3145e7e64c42be7a8c2f7d49

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:452]: Tried on production again, it takes between 2 minutes and 3 minutes. Last time I was about to leave my apartment so I didn't wait until it finished, so I guess last time when I query couch directly maybe it longer than 3 minutes. There is improvement by using elastic search. Is the improvement acceptable? I think maybe yes? Because not every enterprise account would have that amount of mobile workers...

**Comment by jingcheng16** [comment]
@mjriley @biyeun Addressed all feedback and this is ready for review again.

**Comment by jingcheng16** [comment]
@mjriley Thank you for helping me resolving the merge issue. I have updated all my reply with the new SHA and it is ready for review.

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py]: Nice, thanks!

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py:472]: If we're talking minute I think we need to think about caching or turning this into a celery task. I can imagine that multiple requests to this page could cause a disruption in our responsiveness that would be hard to diagnose

File: corehq/apps/reports/standard/deployments.py
Comment (mjriley) [inline corehq/apps/reports/standard/deployments.py:358]: It looks like the prior version of this function only set `last_seen` once, and that line still exists below (where its being set to `last_sub[submission_date]`. Are we setting `last_seen` here only to overwrite it? Should we ignore the `last_seen` return value from this function and rely on the other line? Or can we generalize this logic so that we remove the old line and replace it with the result of this function?

File: corehq/apps/reports/util.py
Comment (mjriley) [inline corehq/apps/reports/util.py:871]: I know most of this was copied from the previous implementation, but there is a bit of duplication for handling the version. Is the following pseudocode better?
```
if ...
    version = last_submission.get('commcare_version')
    ...
elif ...
   version = last_device.get('commcare_version')
   ...

if version:
  version_in_use = format_commcare_version(version) if formatted else version
```

an improvement?

**Review by mjriley — COMMENTED** [review]
Thanks for the changes. I think the initial review is still largely the feeling -- I think the code itself is solid, but I'm concerned about a long-running task blocking our webworkers

File: corehq/apps/reports/util.py
Comment (jingcheng16) [reply corehq/apps/reports/util.py]: Thank you! 34e358b6c09fb54a18f09b7b348a4cb520870498

File: corehq/apps/reports/standard/deployments.py
Comment (jingcheng16) [reply corehq/apps/reports/standard/deployments.py]: Thanks for catching it. This is my oversight. 0d85d21970c4ab93bfffe1300cee233978fb8027
The logic cannot be generalized, because here the `last_seen` will only be last submission time, while the date returned by `get_commcare_version_and_date_from_last_usage` can be fall back to last_device.

File: corehq/apps/builds/utils.py
Comment (mjriley) [inline corehq/apps/builds/utils.py:81]: The use of `memoized` makes me nervous, as my understanding is that the cached value will stick around until it is explicitly cleared, or the machine restarts. Say we created a version, this ran for that version, and then we had to delete the version. This function would still return the old build time, despite the version no longer existing, right? Worse still, because the cache is local to the machine, one webworker might return the old value, while a different webworker, that hadn't previously processed the request, would return the current value. Does it seem safer to perform manual, per-request caching? I.e. create a dictionary and insert values as the request generates them, but throw that dictionary away at the end of the request.

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:464]: Nit: I believe we refer to a domain object as `domain_obj` and the domain name as `domain`

File: corehq/apps/reports/util.py
Comment (mjriley) [inline corehq/apps/reports/util.py:881]: Would it be possible to convert `date_of_use` from a string to a datetime here? It doesn't look like `date_of_use` is referenced by the existing report functions, and this would simplify the logic in `enterprise`, as we could remove the logic that checks whether this is a date or a string

File: corehq/apps/enterprise/static/enterprise/js/project_dashboard.js
Comment (mjriley) [inline corehq/apps/enterprise/static/enterprise/js/project_dashboard.js:273]: @biyeun I did a brief search, but can you recommend any reading on this practice?

**Review by mjriley — COMMENTED** [review]
Great findings and improvements. Before I approve this, I just want to make sure our performance optimizations are worth it and don't have negative side effects. I left commends regarding our use of `memoized` and running the operations in parallel.

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:494]: Seconds and microseconds are removed from the form submission time, not our build. So the result is, for example, if we release a new version `Version 3.3` at 13:10:30, a mobile worker submit a form using `Version 3.2` in 13:10:40, after removing seconds, it is become 13:10:00, the latest version at 13:10:00 is still `Version 3.2`, so this mobile worker won't be considered as using the out-of-date commcare version, although technically he is using out-of-date commcare version. I'm sacrificing a bit accuracy for caching the result of `get_latest_version_at_time`.

File: corehq/apps/enterprise/templates/enterprise/project_dashboard.html
Comment (jingcheng16) [reply corehq/apps/enterprise/templates/enterprise/project_dashboard.html]: 7599675a56c9820eb540ff0802cf236a9e549f4e

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: 7599675a56c9820eb540ff0802cf236a9e549f4e

File: corehq/apps/builds/utils.py
Comment (jingcheng16) [reply corehq/apps/builds/utils.py]: Thank you for noticing it! [I verified in our doc](https://commcare-hq.readthedocs.io/caching_and_memoization.html#lifecycle):
> Memoized on any other function/callable:
The cache lives on the callable, so if it’s globally scoped and never gets garbage collected, neither does the cache


So I'm changing it to a quickcache, and clear it every 10 minutes: 44d9db5

File: corehq/apps/builds/utils.py
Comment (jingcheng16) [reply corehq/apps/builds/utils.py]: Thank you! Sorry this is being extracted into its own function and memoized in a rush. Fixed in bb87820

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py:494]: If we choose to go forward with this approach, I'd want it documented that we're reducing the accuracy of results in order to improve performance. I'm not sure reducing the accuracy of report data for the sake of a quicker summary makes sense. However, can we avoid this entirely now?  It looks like the main bottlenecks for this function were fetching the config and build time, but both of those things have now been moved into their own areas. The config is only fetched once, outside of `get_latest_version_at_time`, and `get_build_time` is now cached. So I don't think we need to do any caching on `get_latest_version_at_time` anymore. At least on staging, our config has 65 items. I don't think it should be expensive to iterate backward through those items until we find a build time less than our target.

File: corehq/apps/builds/utils.py
Comment (mjriley) [reply corehq/apps/builds/utils.py]: Nit: Your version is an improvement, and it hides the exception complexity, but it still does more than is needed in the try/except block. `CommCareBuild` is the statement that can throw, so its slightly better and more explicit to write:
```python
try:
  build = CommCareBuild.get_build(version, latest=True)
except KeyError:
  return None
return build.time if build and build.time else None
```

because it scopes the exception block to _just_ the code that can raise the exception. I also prefer to return an explicit value, rather than letting it implicitly return nothing.

File: corehq/apps/builds/utils.py
Comment (mjriley) [reply corehq/apps/builds/utils.py]: Is the intention 10 minutes? `100 * 60` seems to be 100 minutes, right? I'd still prefer a completely local, request-based caching mechanism, because I think it's more in-line with what you're trying to do, but I'll defer to your judgment here. This code works. The downside is that A) caching is harder to debug, B) quickcache is making redis/network requests, and C) we can still run into that inconsistent state due to a stale cache, but now that stale cache will expire after the timeout.

File: corehq/apps/reports/util.py
Comment (jingcheng16) [reply corehq/apps/reports/util.py:885]: Thank you. I agree with you!! 27aa65f30769b562b0d5f19f1b51fa8f50045ffa

File: corehq/apps/builds/utils.py
Comment (jingcheng16) [reply corehq/apps/builds/utils.py]: 84f1080d10515d9361af2478937078c172128a93

File: corehq/apps/builds/utils.py
Comment (jingcheng16) [reply corehq/apps/builds/utils.py]: Hi Matt, my initial intention is 10 minutes. But now I think 100 minutes is better. I think we will only run into inconsistent state if we release a version then delete it, which I believe this rarely happen. So no need to worry about (A) and (C). By setting `memoize_timeout`, we reduce Redis requests by serving cached values directly from memory if the same process makes repeated calls.

File: corehq/apps/enterprise/enterprise.py
Comment (gherceg) [reply corehq/apps/enterprise/enterprise.py]: Thanks for raising this concern Matt.

> If each of them tries to max out their elasticsearch connections, it seems like we're basically back to running operations in sequence, rather than in parallel?

I'm understanding this as a hypothetical right? Basically saying if we were to use this same technique on every tile, we would end up waiting for I/O elsewhere in the chain? But there is potentially still a world where we employ this technique on tile queries where it is especially useful? We don't have to necessarily max out the connection limit for this worker as I imagine there are still benefits to chunking this by as little as 2 or 3.

And just to reiterate, the larger concern I have is not with this performance change being moot, but it actually being harmful in that it could max out elasticsearch's search pool queue, as Jing noted, resulting in errors.

File: corehq/apps/builds/utils.py
Comment (jingcheng16) [reply corehq/apps/builds/utils.py]: I changed my mind!!! Quickcache added a lot of overhead and resulted in worse performance issue, so I implemented a dictionary to store the cache as you suggested: 83aecc62dbf1d35ee75654fc405b673ad7bc1024

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: Right. It is removed in 83aecc62dbf1d35ee75654fc405b673ad7bc1024

File: corehq/apps/builds/utils.py
Comment (mjriley) [reply corehq/apps/builds/utils.py]: Love it! Daniel did mention to me, however, that the problem with the initial `@memoized` solution wasn't the caching so much as the _global_ caching. Rather than passing the cache in explicitly, and having to write the logic to insert `None` into the cache when an exception is thrown, for example, it might be cleaner to create an object that can limit the scope of the memoization. But entirely optional -- what you implemented is what I had asked for

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py]: Thank you!

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py]: Right, my concern was with adding needless complexity, if these performance improvements didn't turn out to be performance improvements. It sounds like you're concerned that they could lead to system stability issues, which is definitely more pressing. We definitely want to preserve the ability to selectively optimize specific queries, but I think we'd want to specifically call that out here with comments. Otherwise, I think there's a pretty high chance that this code gets copied as an example of how we write tile queries, and it leads to the situation I outlined, where every tile requests max capacity, resulting in no one really getting performance improvements

File: corehq/apps/enterprise/enterprise.py
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:453]: I feel like nested functions generally make the code more confusing. One good use for inner functions is when the inner functions depends on a value from the outer function (i.e. closures). A passable reason would be if the function is extremely simple (a one-liner) and creating an inner function improves readability. I don't think this function meets either of those criteria, though. Nesting it this way means it becomes harder to debug the nested function.

Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py:453]: Oh...I guess it technically is a closure, as it uses the `version_cache`. I guess the choice to nest this function was to avoid having `total_for_domain` take the `version_cache` parameter?

File: corehq/apps/reports/util.py
Comment (mjriley) [inline corehq/apps/reports/util.py:883]: Is this check necessary? I think `date_of_use` will always be a string here.  I think the conditional suggests that sometimes it won't be a string, which is confusing if that isn't true.

File: corehq/apps/enterprise/enterprise.py
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:453]: My initial intention is not avoid having total_for_domain take the version_cache parameter. But I think you just find another reason to next it. :joy: I nest it in total because, otherwise total_for_domain looks like I want to override the total_for_domain in the base class. While I have no intention to override it, I will compute total in a special way in this tile.

File: corehq/apps/builds/utils.py
Comment (mjriley) [inline corehq/apps/builds/utils.py:67]: These lines are now unnecessary with the latest changes, right? We can be assured that `target_time` will be a `datetime` object?

**Review by mjriley — APPROVED** [review]
Thanks for the updates! I think we're at a point where the current code is appropriate for release, so approving.

File: corehq/apps/reports/util.py
Comment (jingcheng16) [reply corehq/apps/reports/util.py]: 82bb7616018d97ba7eec526d941a0595f20cdf53

File: corehq/apps/builds/utils.py
Comment (jingcheng16) [reply corehq/apps/builds/utils.py]: 82bb7616018d97ba7eec526d941a0595f20cdf53

**Review by mjriley — APPROVED** [review]

**Pull request merged**
